From 7e7cf0ab5c70fbacc6092b129b3e1093bc0e9716 Mon Sep 17 00:00:00 2001
From: Eric Sandeen <sandeen@redhat.com>
Date: Thu, 9 Oct 2014 21:00:40 -0400
Subject: [fs] xfs: project id inheritance is a directory only flag

Message-id: <1412888441-4181-47-git-send-email-sandeen@redhat.com>
Patchwork-id: 97328
O-Subject: [RHEL7.1 PATCH 46/47] [fs] xfs: project id inheritance is a directory only flag
Bugzilla: 1145837
RH-Acked-by: Brian Foster <bfoster@redhat.com>
RH-Acked-by: Dave Chinner <dchinner@redhat.com>

Bugzilla: 1145837
Upstream Status: Reviewed on list upstream
Build Info: https://brewweb.devel.redhat.com/taskinfo?taskID=8082426
Tested: locally using xfstests

Thanks,
-Eric

    xfs: project id inheritance is a directory only flag

    From: Dave Chinner <dchinner@redhat.com>

    xfs_set_diflags() allows it to be set on non-directory inodes, and
    this flags errors in xfs_repair. Further, inode allocation allows
    the same directory-only flag to be inherited to non-directories.
    Make sure directory inode flags don't appear on other types of
    inodes.

    This fixes several xfstests scratch fileystem corruption reports
    (e.g. xfs/050) now that xfstests checks scratch filesystems after
    test completion.

    Signed-off-by: Dave Chinner <dchinner@redhat.com>
    Reviewed-by: Brian Foster <bfoster@redhat.com>
Signed-off-by: Eric Sandeen <sandeen@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/xfs/xfs_inode.c b/fs/xfs/xfs_inode.c
index 48ed479..9cbcf14 100644
--- a/fs/xfs/xfs_inode.c
+++ b/fs/xfs/xfs_inode.c
@@ -789,6 +789,8 @@ xfs_ialloc(
      di_flags |= XFS_DIFLAG_EXTSZINHERIT;
      ip->i_d.di_extsize = pip->i_d.di_extsize;
     }
+    if (pip->i_d.di_flags & XFS_DIFLAG_PROJINHERIT)
+     di_flags |= XFS_DIFLAG_PROJINHERIT;
    } else if (S_ISREG(mode)) {
     if (pip->i_d.di_flags & XFS_DIFLAG_RTINHERIT)
      di_flags |= XFS_DIFLAG_REALTIME;
@@ -809,8 +811,6 @@ xfs_ialloc(
    if ((pip->i_d.di_flags & XFS_DIFLAG_NOSYMLINKS) &&
        xfs_inherit_nosymlinks)
     di_flags |= XFS_DIFLAG_NOSYMLINKS;
-   if (pip->i_d.di_flags & XFS_DIFLAG_PROJINHERIT)
-    di_flags |= XFS_DIFLAG_PROJINHERIT;
    if ((pip->i_d.di_flags & XFS_DIFLAG_NODEFRAG) &&
        xfs_inherit_nodefrag)
     di_flags |= XFS_DIFLAG_NODEFRAG;
diff --git a/fs/xfs/xfs_ioctl.c b/fs/xfs/xfs_ioctl.c
index 4d1a479..a58bab5 100644
--- a/fs/xfs/xfs_ioctl.c
+++ b/fs/xfs/xfs_ioctl.c
@@ -997,8 +997,6 @@ xfs_set_diflags(
   di_flags |= XFS_DIFLAG_NOATIME;
  if (xflags & XFS_XFLAG_NODUMP)
   di_flags |= XFS_DIFLAG_NODUMP;
- if (xflags & XFS_XFLAG_PROJINHERIT)
-  di_flags |= XFS_DIFLAG_PROJINHERIT;
  if (xflags & XFS_XFLAG_NODEFRAG)
   di_flags |= XFS_DIFLAG_NODEFRAG;
  if (xflags & XFS_XFLAG_FILESTREAM)
@@ -1010,6 +1008,8 @@ xfs_set_diflags(
    di_flags |= XFS_DIFLAG_NOSYMLINKS;
   if (xflags & XFS_XFLAG_EXTSZINHERIT)
    di_flags |= XFS_DIFLAG_EXTSZINHERIT;
+  if (xflags & XFS_XFLAG_PROJINHERIT)
+   di_flags |= XFS_DIFLAG_PROJINHERIT;
  } else if (S_ISREG(ip->i_d.di_mode)) {
   if (xflags & XFS_XFLAG_REALTIME)
    di_flags |= XFS_DIFLAG_REALTIME;
-- 
1.7.1